src/rgw/rgw_pubsub.cc

```diff
@@ -0,0 +1,495 @@
+    return -ENOENT;
+  }
+
+  auto& t = iter->second;
+
+  rgw_pubsub_sub_config sub_conf;
+
+  sub_conf.user = ps->user;
+  sub_conf.name = sub;
+  sub_conf.topic = topic;
+  sub_conf.dest = dest;
+
+  t.subs.insert(sub);
+
+  ret = ps->write_user_topics(topics, &user_objv_tracker);
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: failed to write topics info: ret=" << ret << dendl;
+    return ret;
+  }
+
+  ret = write_sub(sub_conf, nullptr);
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: failed to write subscription info: ret=" << ret << dendl;
+    return ret;
+  }
+  return 0;
+}
+
+int RGWUserPubSub::Sub::unsubscribe(const string& _topic)
+{
+  string topic = _topic;
+  RGWObjVersionTracker sobjv_tracker;
+  RGWRados *store = ps->store;
+
+  if (topic.empty()) {
+    rgw_pubsub_sub_config sub_conf;
+    int ret = read_sub(&sub_conf, &sobjv_tracker);
+    if (ret < 0) {
+      ldout(store->ctx(), 0) << "ERROR: failed to read subscription info: ret=" << ret << dendl;
+      return ret;
+    }
+    topic = sub_conf.topic;
+  }
+
+  RGWObjVersionTracker objv_tracker;
+  rgw_pubsub_user_topics topics;
+
+  int ret = ps->read_user_topics(&topics, &objv_tracker);
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: failed to read topics info: ret=" << ret << dendl;
+  }
+
+  if (ret >= 0) {
+    auto iter = topics.topics.find(topic);
+    if (iter != topics.topics.end()) {
+      auto& t = iter->second;
+
+      t.subs.erase(sub);
+
+      ret = ps->write_user_topics(topics, &objv_tracker);
+      if (ret < 0) {
+        ldout(store->ctx(), 0) << "ERROR: failed to write topics info: ret=" << ret << dendl;
+        return ret;
+      }
+    }
+  }
+
+  ret = remove_sub(&sobjv_tracker);
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: failed to delete subscription info: ret=" << ret << dendl;
+    return ret;
+  }
+  return 0;
+}
+
+void RGWUserPubSub::Sub::list_events_result::dump(Formatter *f) const
+{
+  encode_json("next_marker", next_marker, f);
+  encode_json("is_truncated", is_truncated, f);
+
+  Formatter::ArraySection s(*f, "events");
+  for (auto& event : events) {
+    encode_json("event", event, f);
+  }
+}
+
+int RGWUserPubSub::Sub::list_events(const string& marker, int max_events,
+                               list_events_result *result)
+{
+  RGWRados *store = ps->store;
+  rgw_pubsub_sub_config sub_conf;
+  int ret = get_conf(&sub_conf);
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: failed to read sub config: ret=" << ret << dendl;
+    return ret;
+  }
+
+  RGWBucketInfo bucket_info;
+  string tenant;
+  RGWObjectCtx obj_ctx(store);
+  ret = store->get_bucket_info(obj_ctx, tenant, sub_conf.dest.bucket_name, bucket_info, nullptr, nullptr);
+  if (ret < 0) {
+    ldout(store->ctx(), 0) << "ERROR: failed to read bucket info for events bucket: bucket=" << sub_conf.dest.bucket_name << " ret=" << ret << dendl;
```

Comment:
i remember this being a source of confusion in testing, because the bucket isn't created until it's populated with events. would it make sense to return an empty list on ENOENT here? or maybe handle ENOENT gracefully in rgw_admin.cc for OPT_PUBSUB_SUB_PULL?

Reply:
yeah, empty list makes sense